benchmark rpc ps

gcramer23 opened the pull request:
DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57454 benchmark rpc ps**

Differential Revision: [D28296175](https://our.internmc.facebook.com/intern/diff/D28296175)

gcramer23 committed:
benchmark rpc ps

[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:
```diff
@@ -0,0 +1,25 @@
+
+    def elapsed_time(self, rank):
+        self.start.synchronize()
+        self.end.synchronize()
```
This means if this is called before the CUDA ops are done, it will be perf impacts. Shall we make the synchronize API explicit? I mean 


```python
    def elapsed_time(self, rank):
        if not self.end.query():
            raise RuntimeError("......")
        return self.start.elapsed_time(self.end)

    def synchronize():
        ....
```

gcramer23 replied:
okay ill add the synchronize API.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:
```diff
@@ -0,0 +1,25 @@
+    def elapsed_time(self, rank):
+        self.start.synchronize()
+        self.end.synchronize()
+        return self.start.elapsed_time(self.end)
```
IIRC, this will fail if `self.start` and `self.end` are on different devices?

gcramer23 replied:
Yes, it does fail. The API calls to `record_start` and `record_end` are from the logger. The logger is initialized with a device rank. So it shouldn't be a problem if a user is inheriting from a base class since they can't access the logger because of name mangling.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:
```diff
@@ -0,0 +1,25 @@
+        self.start = None
+        self.end = None
+
+    def record_start(self, rank):
```
Do we need to pass in rank here, or can it be specified in the ctor? Besides, passing it here means that this is not overriding the abstract `record_start` in `MetricBase`?

gcramer23 replied:
Well python does not call the parent method if a subclass has the method name. 

```
class A:

    def __init__(self):
        self.name = "A"

    def get_name(self):
        return self.name

class B(A):

    def __init__(self):
        super().__init__()
        self.name = "B"

    def get_name(self, name):
        return name

if __name__ == "__main__":
    b = B()
    # print(b.get_name("garrett"))
    print(b.get_name())
```

gcramer23 replied:
> can it be specified in the ctor

I think this is a better approach because it also deals with the issue of the metric recording on different devices. https://github.com/pytorch/pytorch/pull/57454#discussion_r625474043

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricBase.py:
```diff
@@ -0,0 +1,29 @@
+    def get_name(self):
+        return self.name
+
+    def get_start(self):
```
API not used.

gcramer23 replied:
okay will remove it.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+    Each key has a list of elapsed times.
+
+     Args:
+        self: the instance of the class
```
API doc usually don't include self, because users would use it as `instance. get_processed_metrics()`.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+    def get_metrics(self):
+        return self.metrics
+
+    r"""Method that processes the Metrics. 
```
If this mean to be an API doc, it should be placed inside the function implementation. See the following example.

https://github.com/pytorch/pytorch/blob/1d3a9bff3c3256bd8aec4ed066119eb32883715c/torch/distributed/rpc/api.py#L454-L460

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+    CUDA = "cuda"
+    CPU = "cpu"
+
+    def __init__(self, rank, metric_class, overwrite_metrics=False):
```
If `overwrite_metrics=True` was never used in this PR. Shall we remove it from this one and add it back when needed in the future?

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+
+    def __init__(self, rank, metric_class, overwrite_metrics=False):
+        self.rank = rank
+        self.metric_class = metric_class.lower()
```
this means we don't expect to use the same logger to log both CPU and CUDA metrics?

gcramer23 replied:
I was thinking about this. CPUMetrics are not used in the experiments that I am implementing, but I can update the logger to specify logging CUDA or CPU.

```python

    def cpu_record_start(self, type, key, name):
        ...

    def cuda_record_start(self, type, key, name):
        ...            

```

I'll update the existing wrapper in the base classes and add a CPU wrapper.

gcramer23 replied:
I ended up adding a CUDA parameter to indicate if the metric should be CUDA or CPU.
```python
    def record_start(self, type, key, name, cuda=True):
        self.__metrics_logger.record_start(
            type,
            key,
            name,
            cuda
        )
 ```

mrshenli commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+
+def get_ps_metrics(configurations, ps_configurations, parameter_server_rrefs):
+    rank_metrics = []
+    if configurations.parameter_server_count > 0:
+        ps = get_benchmark_ps_map()[str(ps_configurations["ps_class"])]
+        for rank, ps_rref in parameter_server_rrefs.items():
+            metrics = rpc.rpc_sync(
+                ps_rref.owner(),
+                ps.get_metrics_rpc,
+                args=(ps_rref,)
+            )
+            rank_metrics.append([rank, metrics])
```
IIUC, this is not used in this PR? If yes, let's remove and add it when we need it.

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+    return rank_mod_ps_count + configurations.trainer_count
+
+
+def print_metrics(metrics_printer, rank_metrics_list, name):
```
curious, why do we need an additional helper function instead of putting this to the metrics printer class implementation?

gcramer23 replied:
Ill move it to the metrics printer class.

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+        if metric_type not in self.metrics:
+            self.metrics[metric_type] = {}
+
+        if self.metric_class == self.CPU:
+            self.metrics[metric_type][key] = CPUMetric(metric_name)
+        elif self.metric_class == self.CUDA:
+            self.metrics[metric_type][key] = CUDAMetric(metric_name)
```
If this the intention, I guess you can allow the user to directly pass in a metric class (not string) to the ctor. I.e.:

```python
def __init__(self, rank, metric_class):
    self.metric_class = metric_class

def add_metric(...):
    self.metrics[type][key] = self.metric_class(name)
```

And users can call the ctor with 

```python
logger = MetricsLogger(0, CUDAMetric)
```

gcramer23 replied:
I think just allowing the logger to log both is a better solution. https://github.com/pytorch/pytorch/pull/57454#discussion_r626144963

mrshenli commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+        self.overwrite_metrics = overwrite_metrics
+        self.metrics = {}
+
+    def add_metric(self, metric_type, key, metric_name):
```
the API name already contains a metric, we can simplify the args to type, key, name

gcramer23 replied:
okay.

mrshenli commented on benchmarks/distributed/rpc/ps/trainers/TrainerBase.py:
```diff
@@ -0,0 +1,92 @@
+        return
+
+    def record_start(self, metric_type, key, metric_name):
+        self.__metrics_logger.add_metric(
```
curious, is there any reason that need a different name. The call trace is like:

record_start (TrainerBase) -> add_metric (MetricLogger) -> record_start(Metric)

Will there be any issue if we all stay with the same name?

gcramer23 replied:
no but adding the `cuda_record_start` and `cpu_record_start` the call trace will be

cuda_record_start (TrainerBase) -> cuda_record_start (MetricLogger) -> record_start(CUDAMetric)
cpu_record_start (TrainerBase) -> cpu_record_start (MetricLogger) -> record_start(CPUMetric)

gcramer23 replied:
call trace is now record_start (TrainerBase) -> record_start (MetricLogger) -> record_start(Metric)

mrshenli commented on benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py:
```diff
@@ -0,0 +1,107 @@
+        def epoch_key(epoch, index):
+            return "{},{}".format(epoch, index)
+
+        for epoch in range(self.epochs):
+            for index, batch in enumerate(data):
+                hook_state.next_batch_state()
+                input, target = batch[0], batch[1]
+
+                self.record_batch_start(epoch_key(epoch, index))
+
+                optimizer.zero_grad()
+
+                self.record_forward_start(epoch_key(epoch, index))
+
+                out = ddp_model(input)
+
+                self.record_forward_end(epoch_key(epoch, index))
+
+                loss = criterion(out, target)
+
+                self.record_backward_start(epoch_key(epoch, index))
+
+                loss.backward()
+
+                self.record_backward_end(epoch_key(epoch, index))
+
+                optimizer.step()
+
+                self.record_batch_end(epoch_key(epoch, index))
+
+        torch.cuda.synchronize(self.rank)
```
I wonder if this training loop can be moved to the TrainerBase class?

mrshenli replied:
Would I be correct if I assume different trainers only need different hook function and hook state, and all other code can be moved to TrainerBase?

gcramer23 replied:
It is hard to determine that all training loops will be identical. For example, maybe I want to call backward multiple times in my training loop.  

All trainers that inherit from DdpTrainerBase will have to implement their hook function and hook state. If a trainer is implemented that does not inherit from DdpTrainerBase, it does not need to implement a hook function and hook state. It will potentially have a different training loop that is not based on a communication hook.

I don't think the training loop should be in the base class.

H-Huang commented on benchmarks/distributed/rpc/ps/BenchmarkConfigurations.py:
```diff
@@ -0,0 +1,31 @@
+class BenchmarkConfigurations:
```
python dataclass could simplify this class (https://docs.python.org/3/library/dataclasses.html)

gcramer23 replied:
okay I will use it.

H-Huang commented on benchmarks/distributed/rpc/ps/README.md:
```diff
@@ -0,0 +1,62 @@
+# RPC PS Benchmark
```
nit: Change folder name from ps -> parameter_server? `.../rpc/parameter_server` sounds more intuitive than `.../rpc/ps` to me

gcramer23 replied:
okay.

H-Huang commented on benchmarks/distributed/rpc/ps/README.md:
```diff
@@ -0,0 +1,62 @@
+# RPC PS Benchmark
+
+## How to add your experiment
```
This can be done at a later time. Perhaps creating a placeholder experiment and a quickstart of what to change in the placeholder experiment could better illustrate the steps needed

gcramer23 replied:
Okay, Ill make a note to do this later.

H-Huang commented on benchmarks/distributed/rpc/ps/bash_experiment_scripts/ddp_nccl_allreduce.sh:
```diff
@@ -0,0 +1,14 @@
+
+# moves to directory and runs the benchmark with the configurations selected
+cd "$(dirname $(dirname "$0"))"
+source ./bash_experiment_scripts/helper_functions.sh
```
may want to note here or in the README that you need slurm to run this script

gcramer23 replied:
ill add it to the script.

H-Huang commented on benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:
```diff
@@ -0,0 +1,92 @@
+
+    def add_metric(self, metric_type, key, metric_name):
+        if metric_type in self.metrics and key in self.metrics[metric_type] and not self.overwrite_metrics:
+            raise Exception("metric_type={} with key={} already exists".format(metric_type, key))
```
can use f-strings here and below as well

gcramer23 replied:
okay.

H-Huang commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+
+def get_name(rank, trainer_count, parameter_server_count):
+    if rank < trainer_count:
+        return "trainer{}".format(rank)
```
nit: use f-strings instead of .format. Same comment for rest of formatted strings in file

gcramer23 replied:
okay.

H-Huang commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+
+def get_parameter_server_rank(rank, configurations):
+    if configurations.parameter_server_count < 1:
+        return None
```
Should we throw an error here instead?

gcramer23 replied:
Ill remove those two lines not needed.

H-Huang commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+    parser.add_argument(
+        "--bconfig_id",
+        type=str,
+        default="1",
```
you can add the `help` parameter to provide information about each argument

gcramer23 replied:
okay.

H-Huang commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+def load_configurations(args):
+    benchmark_config_file = "configurations/benchmark_configurations.json"
+    benchmark_config = json.load(
+        open(
```
Don't we need a context manager when opening files? (e.g. `with open(file, "r") as f: ...` ), or does json.load handle that for you?

gcramer23 replied:
No I use json to parse a file object.
```
f = open(file, "r")
json.load(f)
```

H-Huang replied:
Then we should update this to also close the file right? It is always best practice to close after opening to ensure resources (e.g. file descriptor) are released correctly. Your example above would be changed to:
```python
with open(file, "r") as f:
    json.load(f)
```

gcramer23 replied:
yes forgot to close the file. Ill use close it.

gcramer23 commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+    rpc_backend_options = TensorPipeRpcBackendOptions()
+    rpc_backend_options.init_method = configurations.rpc_init_method
+
+    # trainers = [0, trainer_count)
+    # parameter_servers = [trainer_count, trainer_count + parameter_server_count)
+    # master = [trainer_count + parameter_server_count, trainer_count + parameter_server_count]
```
okay.

gcramer23 commented on benchmarks/distributed/rpc/ps/metrics/CPUMetric.py:
```diff
@@ -0,0 +1,19 @@
+    def record_end(self, rank):
+        self.end = time.time()
+
+    def elapsed_time(self):
```
Ill add a RuntimeError.

gcramer23 commented on benchmarks/distributed/rpc/ps/models/DummyModel.py:
```diff
@@ -0,0 +1,22 @@
+class DummyModel(nn.Module):
+    def __init__(
+        self,
+        num_embeddings=10,
```
Ill remove the values.

gcramer23 commented on benchmarks/distributed/rpc/ps/metrics/ProcessedMetricsPrinter.py:
```diff
@@ -0,0 +1,44 @@
+
+class ProcessedMetricsPrinter:
+
+    def combine_processed_metrics(self, processed_metrics_list):
```
okay.

gcramer23 commented on benchmarks/distributed/rpc/ps/launcher.py:
```diff
@@ -0,0 +1,372 @@
+    return model_class(**model_config)
+
+
+def load_model(args):
+    model_config_file = "configurations/model_configurations.json"
+    model_config = json.load(
+        open(
+            os.path.join(
+                Path(__file__).parent, model_config_file
+            ),
+            "r"
+        )
+    )[args.mconfig_id]
+    return get_model(model_config["model_class"], model_config["configurations"])
```
There are some slight variations. I have added a helper function for loading the JSON.

gcramer23 commented on benchmarks/distributed/rpc/ps/trainers/TrainerBase.py:
```diff
@@ -0,0 +1,92 @@
+        )
+
+    @staticmethod
+    def methodmetric(metric_name, metric_type="method_metric"):
```
Not in this pr, but I think this is a good decorator to have in the base class.

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -0,0 +1,320 @@
+def run_master(rank, model, data, configurations, trainer_configurations, ps_configurations, rpc_backend_options):
+    world_size = configurations.trainer_count + configurations.parameter_server_count + 1
+    rpc.init_rpc(
+        get_name(
```
nit: will it be better if `get_name` just take the `configurations` as an argument and then retrieve whatever fields it needs? It will help make the call site less verbose.

mrshenli replied:
also in case there will be additional roles in the future, we don't need to update all call sites.

gcramer23 replied:
okay updated the get_name method.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -0,0 +1,320 @@
+    metrics_printer.print_metrics("trainer", rank_metrics_list)
+
+
+def run_benchmark(rank, model, data, configurations, trainer_configurations, ps_configurations):
```
curious, any reason that configurations, trainer_configurations, ps_configurations needs to be three different objects. The files can be separate, but can they all load into one `config` object and then functions using that config can just retrieve whatever they need?

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -0,0 +1,320 @@
+
+
+def load_trainer_configurations(args):
+    trainer_config_file = "configurations/trainer_configurations.json"
```
shall we change the hard-coded path and let users to pass in a file path using command line args?

mrshenli replied:
this applies to other configurations as well.

gcramer23 replied:
okay updated.

mrshenli commented on benchmarks/distributed/rpc/parameter_server/launcher.py:
```diff
@@ -0,0 +1,320 @@
+def main():
+    parser = argparse.ArgumentParser(description="RPC PS Benchmark")
+    parser.add_argument(
+        "--bconfig_id",
```
after changing this to take a file path, let's also update the config names. We can call then --benchmark, --data, --model, --server, --trainer, for readability.

mrshenli reviewed (approved):
I left some minor comments. Stamp to unblock. Please consider addressing the minor comments before landing. I will create Github issues if I discover sth after landing. Thanks for working on this!

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```




[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28296175).

gcramer23 committed:
Update on "benchmark rpc ps"

DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```


Differential Revision: [D28296175](https://our.internmc.facebook.com/intern/diff/D28296175)

[ghstack-poisoned]
